pkg/hub/controllers/cluster/deregister.go

```diff
@@ -0,0 +1,302 @@
+		log.Error(err, "error updating status of deregistration on the controller")
+		return err
+	}
+	// Create a service account.
+	if err := r.MeshClient.Create(ctx, constructServiceAccount(), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("service account already exists", "sa", serviceAccountName)
+		} else {
+			log.Error(err, "unable to create service account")
+			return err
+		}
+	}
+	// get operator clusterrole
+	operatorClusterRole, err := r.getOperatorClusterRole(ctx)
+	if err != nil {
+		log.Error(err, "unable to fetch operator clusterrole")
+		return err
+	}
+	// Create a cluster role.
+	if err := r.MeshClient.Create(ctx, constructClusterRole(operatorClusterRole), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster role already exists", "clusterrole", clusterRoleName)
+		} else {
+			log.Error(err, "unable to create cluster role")
+			return err
+		}
+	}
+	// Bind the service account with the cluster role.
+	if err := r.MeshClient.Create(ctx, constructClusterRoleBinding(), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster rolebinding already exists", "clusterrolebinding", clusterRoleBindingName)
+		} else {
+			log.Error(err, "unable to create cluster rolebinding")
+			return err
+		}
+	}
+	// get configmap data
+	data, err := getConfigmapData()
+	if err != nil {
+		log.Error(err, "unable to fetch configmap data")
+		return err
+	}
+	// Set up a configuration map which contains the script for the job.
+	if err := r.MeshClient.Create(ctx, constructConfigMap(data), &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster configmap already exists", "configmap", clusterDeregisterConfigMap)
+		} else {
+			log.Error(err, "unable to create cluster configmap")
+			return err
+		}
+	}
+
+	// get the job
+	jobRef := types.NamespacedName{
+		Name:      deregisterJobName,
+		Namespace: ControlPlaneNamespace,
+	}
+	deregisterJob := constructJobForClusterDeregister()
+	if err := r.MeshClient.Get(ctx, jobRef, deregisterJob); err == nil {
+		// Job is already present we need to delete it
+		err = r.MeshClient.Delete(ctx, deregisterJob)
+		if err != nil {
+			log.Error(err, "error while deleting job object", "job", jobRef.Name)
+			return err
+		}
+	}
+
+	// Generate a job for deregistration with the configuration map mounted as a volume.
+	if err := r.MeshClient.Create(ctx, deregisterJob, &client.CreateOptions{}); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("cluster deregister job already exists", "job", clusterDeregisterConfigMap)
+		} else {
+			log.Error(err, "unable to create job for cluster deregister")
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) updateRegistrationStatusAndDeregisterInProgress(ctx context.Context, cluster *hubv1alpha1.Cluster, status hubv1alpha1.RegistrationStatus, isDeregisterInProgress *bool) error {
```

Comment:
Not sure if i understand the need for the "isDeregisterInProgress *bool" param. A couple of questions:

Why does it have to be a pointer? 
From the code, it is only set to false in the handleDeletion func in the other file. Dont see the flag being set to true. When is it set to true?
Why cant the hubv1alpha1.RegistrationStatusDeregisterInProgress be used to check the status instead of having this separate flag?

Reply:
It will impact controller side if we remove isDeregisterInProgress bool. @priyank-upadhyay could please give a little explanation why it is needed.

Reply:
And pointer because we don't want to set it true, controller will do it and to use same function to update RegistrationStatus and IsDeregisterInProgress we wrote it like that @bharath-avesha